### ISSUE TITLE
SocketMode support

### ISSUE THREAD

**Issue by hughesmk** [soxtoby/SlackNet#67]
Simon, Slack recently announced Socket Mode which basically enables those of us behind firewalls that will not be opened to use Events and Web API across the websocket.  It supposedly provides the benefits of all three APIs through the web socket.  

Slack has released support for in their Bolt SDK and it sounds like the feature could be available in about 60 days.

I wanted to see if this was on your radar, and offer to help with testing.  I am not a coder of your skill level but can help with testing, etc if you are interested, and am happy to share anything I cobble together in the meantime.

**Comment by soxtoby** [comment]
Thanks for bringing this to my attention, @hughesmk. I've been wondering what to do with SlackNet.Bot, since Slack has basically deprecated the RTM API that it's based on, and this seems to be the answer.

I'll definitely be looking at this as soon as I can, and I'll appreciate any feedback when the time comes, since Slack's documentation often misses details that you can only find by trying things out.

**Comment by hughesmk** [comment]
I have the code logging in successfully in socket mode. I need to work on handling the events next.  I have some documentation that I can share.  On Nov 6, 2020 6:02 PM, Simon Oxtoby <notifications@github.com> wrote:
Thanks for bringing this to my attention, @hughesmk. I've been wondering what to do with SlackNet.Bot, since Slack has basically deprecated the RTM API that it's based on, and this seems to be the answer.
I'll definitely be looking at this as soon as I can, and I'll appreciate any feedback when the time comes, since Slack's documentation often misses details that you can only find by trying things out.

**Comment by hughesmk** [comment]
I have pretty much gotten SlackNet working with SocketMode.  I am sure that I have not yet found all the potential issues but at present things appear to be working fairly well.   The one outstanding item that I have open with Slack is that the bot never goes online as it would in RTM and toggling the always online option as you would with an events api bot takes an hour or more to be recognized.  I have suggested that they provide an RTM type solution.

**Comment by robbiebihun** [comment]
Is there a branch with this change that one could review?

**Comment by hughesmk** [comment]
I will be honest, I am unsure how to do that here on github, but am happy.eager to do so if someone can point me the right direction.  I know that I probably have done a few things that need to be done differently long term, but it has been working perfectly for a few weeks now.

**Comment by hughesmk** [comment]
It appears that Slack has officially launched their Socket mode support.

https://api.slack.com/apis/connections/socket

**Comment by soxtoby** [comment]
👍 I started looking at this yesterday. The basic mechanism is pretty simple, which is great, but it's going to require a large refactor of the existing event handling code. One of the main use-cases for socket mode is testing an app before it's published online, so I want to make sure it's easy to switch between socket mode and receiving HTTP requests.

**Comment by hughesmk** [comment]
Wanted to circle back and see if you wanted any help testing or if there was any assistance that I can offer.

**Comment by soxtoby** [comment]
Sorry, been meaning to post an update on this but keep getting side-tracked. I _have_ been making steady progress on this - the large refactor I mentioned above is complete, and I just need to hook the socket mode client up to the event handling and verify that it works.

I'll hopefully push something up either this weekend or next, at which point I'd love for people to give it a go and put it through its paces. In order to get this up as soon as I can, it'll probably be missing ASP.NET integration and multi-connection support, with those being implemented in the following weeks.

Thanks for your patience; I hope to make the wait worth it 🙂

**Comment by soxtoby** [comment]
Socket mode support is now available on the [socketmode branch](https://github.com/soxtoby/SlackNet/tree/socketmode), and on NuGet as [version 0.7.13-socketmode.1](https://www.nuget.org/packages/SlackNet/0.7.13-socketmode.1).

## TL;DR

See the [SlackNet.SocketModeExample](https://github.com/soxtoby/SlackNet/tree/socketmode/SlackNet.SocketModeExample) project for example usage. See [Using a DI container](#using-a-di-container) below for example integration code. With a DI container, handler registration is the same as in [SlackNet.EventsExample](https://github.com/soxtoby/SlackNet/blob/socketmode/SlackNet.EventsExample/Startup.cs).

## Design considerations

The socket mode client performs essentially the same function as the ASP.NET integration, so I've tried to keep the API as close as possible. The end goal is to be able to switch the ASP.NET integration between socket mode and normal request handling with a single option, to make developing an app locally as simple as possible.

The socket mode client needs to work without ASP.NET of course, and this is where things get a little tricky, since I was relying on ASP.NET's built-in dependency injection for handler configuration. My solution was to introduce a standardised way of configuring SlackNet services that can be used on its own, or by piggy-backing off a DI container for more power. I've provided integration packages for Autofac, Microsoft.Extensions.DependencyInjection, and SimpleInjector; these are the libraries I'm most familiar with, but I'm happy to add more integrations if needed.

## Using the built-in SlackServiceFactory
```c#
var slackServices = new SlackServiceFactory();
slackServices
    .UseAppLevelToken("<app-level OAuth token required for socket mode>")
    .RegisterEventHandler(myEventHandler)
    .RegisterBlockActionHandler(myBlockActionHandler);
var client = slackService.GetSocketModeClient();
await client.Connect();
```

`SlackServiceFactory` provides registration methods for all the different types of handlers, as well as the ability to replace specific services with custom implementations, should the need arise.

It also provides methods to retrieve any of the SlackNet services, most notably `GetApiClient` and `GetSocketModeClient`.

## The socket mode client
Similar to the RTM client, once connected, the socket mode client will reconnect automatically, until the client is disposed, or socket mode is disabled.

Each message received is treated as a "request", equivalent to receiving a request through the ASP.NET integration. Responses are returned as acknowledgements through the websocket connection.

The socket mode client seems to be working fine at the moment, but there's a few issues I still need to address:
 - Message retries are not being handled. If you take too long to respond to a message, then you'll probably end up handling the same message again.
 - No multi-connection support. The connection to Slack will periodically need to reconnect, and messages sent during this period may be dropped.
 - No way to handle disconnects.

## Using a DI container
The built-in `SlackServiceFactory` requires services and handlers to be constructed manually, but with a DI container, they can be registered with just a type, leaving construction to the container. Other services registered in the container can be injected as you'd expect, and services and handlers are given appropriate lifestyles.

Since the configuration has been standardised, it's pretty much the same for any container, but here's what it looks like for the provided integrations&hellip;

### Autofac
Using the `SlackNet.Autofac` package:

```c#
var containerBuilder = new ContainerBuilder();
containerBuilder.AddSlackNet(c => c
    .UseAppLevelToken("<app-level OAuth token>")
    .RegisterEventHandler<MyEventHandler>()
    .RegisterBlockActionHandler<MyBlockActionHandler>());
var container = containerBuilder.Build();
var client = container.Resolve<ISlackSocketModeClient>();
await client.Connect();
```

### Microsoft.Extensions.DependencyInjection
Using the `SlackNet.Extensions.DependencyInjection` package:

```c#
var serviceCollection = new ServiceCollection();
serviceCollection.AddSlackNet(c => c
    .UseAppLevelToken("<app-level OAuth token>")
    .RegisterEventHandler<MyEventHandler>()
    .RegisterBlockActionHandler<MyBlockActionHandler>());
var serviceProvider = serviceCollection.BuildServiceProvider();
var client = serviceProvider.GetRequiredService<ISlackSocketModeClient>();
await client.Connect();
```

### SimpleInjector
Using the `SlackNet.SimpleInjector` package:

```c#
var container = new Container {
    // AsyncScopedLifestyle is used for scoping handlers
    Options = { DefaultScopedLifestyle = new AsyncScopedLifestyle() }
};
container.AddSlackNet(c => c
    .UseAppLevelToken("<app-level OAuth token>")
    .RegisterEventHandler<MyEventHandler>()
    .RegisterBlockActionHandler<MyBlockActionHandler>());
var client = container.GetInstance<ISlackSocketModeClient>();
await client.Connect();
```

## Final thoughts
While I do want to finish this feature off and release it properly as soon as possible, I'm definitely open to suggestions on API improvements. I'm worried that using a factory to create the socket mode client may be too counter-intuitive, although I suppose it's fairly straightforward once you see an example. I'm also not entirely happy with the DI container integration code: as much as I tried to simplify it, there's still quite a bit involved in each integration.

**Comment by eschgi** [comment]
Hi,

I have a question, the SlackNet.Bot library currently uses RTM what was deprecated or not?
When I want to build now a new bot I should use the SocketModeClient directly or?

Thank you.

Kind regards
Stefan

**Comment by soxtoby** [comment]
Hi @eschgi, you're right that the RTM client is on its way out. I may move SlackNet.Bot over to the new socket mode client at some point, but for now I'd recommend either the ASP.NET integration, or the socket mode client. Be aware though that the socket mode client is still in preview, so if you start using it now, there will probably be some (hopefully only minor) breaking changes when it's released properly.

**Comment by soxtoby** [comment]
I've just pushed up version [0.8.2-socketmode.2](https://www.nuget.org/packages/SlackNet/0.8.2-socketmode.2); it includes multi-connection support and completed ASP.NET integration. At this point I think socket mode support is complete, so I'll release it in a couple weeks, once I've updated the documentation and if no one hits any major issues.

### Multi-connection support
The socket mode client will open 2 connections, with a 10 second delay between them so they don't expire at the same time. This is currently configurable only if you construct the client yourself, but I'll make the configuration easier before release.

### ASP.NET integration
You can switch the ASP.NET integration over to socket mode by configuring an app-level token with `services.AddSlackNet(c => c.UseAppLevelToken("<your token here>"))` and enabling socket mode with `app.UseSlackNet(c => c.UseSocketMode())`. The registration API has been completely replaced, but shouldn't require any changes since the new API matches the old one.

### More registration options
Handlers can be registered with callbacks that receive the `SlackRequestContext`, which provides access to all the registered services and handlers.

## Breaking Changes
- `IEventsObservables` has been removed. It's simple enough to re-implement if anyone needs it, but it doesn't fit the rest of the API.
- `SlackNet.AspNetCore` now depends on `SlackNet.Extensions.DependencyInjection`.
- `SlackServiceFactory` has been renamed to `SlackServiceBuilder`.
- `IHandlerIndex.TryGetValue` has been renamed to `TryGetHandler`.
- `ISlackServiceFactory` has been renamed to `ISlackServiceProvider`.
- A handful of changes to public types related to the new handler API that hopefully shouldn't affect anyone, since there was no reason to use these types directly.

**Comment by hughesmk** [comment]
This is awesome!  I am working through porting all of my logic from RTM and Slacknet.Bot to socket mode.  A few things that I am tripping on are:

The SlackBot class has an initializer for RTM but not for SocketMode, I am guessing that we will just move off the SlackBot class as we make this transition.

It looks like you took pains to make it easier to use custom http and websocket factories to allow for us folks who need to use proxies, etc.  I have not yet figured out quite the right syntax yet though to make it work, I am using .UseWebSocketFactory(proxySupportingWebSocketFactory)  but of course it is unhappy.   I will take another swing at it later tonight to see if I can decipher what I am doing wrong.

**Comment by s0nyguy** [comment]
Hey, 

I'm testing 0.8.2-socketmode.2 and hit a small issue. I open a modal using the code

    _slack.Views.Open(triggerId, new ModalViewDefinition)..........
    
This works just fine, but on clicking submit button on the modal, I'm not receiving any payload. I've registered a ViewSubmissionHandler with the correct callback id (matching the modal).  

    slackServices.RegisterViewSubmissionHandler(ViewSubmissionHandler.ModalCallbackId, ctx => new ViewSubmissionHandler(ctx.ServiceProvider.GetApiClient()));

The same is true for HandleClose(), it is not fired when closing the modal despite having NotifyOnClose set to true.

Is there anything I'm missing, or could this be a bug?

Thanks!

**Comment by s0nyguy** [comment]
Found the issue:

File: SlackSocketModeClient.cs
Line: 105
Method: HandleSpecifiRequest()
Issue:

    interaction.Payload.TryGetValue("response_url", out)

Should be

    interaction.Payload.TryGetValue("response_urls", out)

Note the plural 'urls'. After I updated this my ViewSubmissionHandler started working as expected.

**Comment by soxtoby** [comment]
Thanks for finding that issue, @s0nyguy. Turns out that whole check needed a rethink. I've pushed up version [0.8.2-socketmode.3](https://www.nuget.org/packages/SlackNet/0.8.2-socketmode.3) with a fix.

This version also allows passing the multi-connection configuration into the client's `Connect` method, and fixes a couple other issues I'd noticed.

**Comment by s0nyguy** [comment]
Happy to help, it's a fantastic library so thanks for all your hard work! I'll pull 0.8.2-socketmode.3 and give that a go.

**Comment by soxtoby** [comment]
Socket mode support is now available in v0.9.0 🎉

Thankyou to everyone who tried it out during development, and thankyou for your patience; this was a much larger task than I anticipated, but I'm pretty happy with how it turned out.

As with any large change, there's bound to be at least some minor bugs, so please raise new issues for anything you come across.
